### ISSUE TITLE
[deeplab] Logits remain original shape 21 despite last_layers_contain_logits_only flag=false

### ISSUE THREAD

**Issue by kayvonkhosrowpour** [tensorflow/models#6559]
### System information
- **What is the top-level directory of the model you are using**: deeplab
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10 Enterprise
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**:  1.12.0
- **Bazel version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: 9
- **GPU model and memory**: NVIDIA Quadro M4000, 8GB
- **Exact command to reproduce**: Does not apply

Please note that I am referring to what I believe to be the same issue that is unresolved (yet closed) in #5040 . This is my first issue but my problem is quite specific and that issue does not provide a working solution and I cannot seem to find the answer.

### Describe the problem
I am attempting to train on my own dataset, called "ade20k_PRDL3_v1", which is a modified version of ade20k. The dataset has only 4 classes and all of the images are the same size. The changes I've made to the source code as well as the training script that I've created seem to result in a training that completes. However, I am unable to evaluate/visualize the results because the logits appear to have the same shape (21 from pascal) instead of the desired shape (4 from ade20k_PRDL3_v1) when running vis.py. As you can see below, I set `initialize_last_layer=false` and `last_layers_contain_logits_only=true` as indicated by several sources.

### Source code / logs
Much of what I changed/created was from [this guide](http://hellodfan.com/2018/07/06/DeepLabv3-with-own-dataset/) but I reproduced the code here...

#### In utils > get_colormap.py:
I added the __PRDL3_V1 max entries:
```
_DATASET_MAX_ENTRIES = {
    _ADE20K: 151,
    _CITYSCAPES: 256,
    _MAPILLARY_VISTAS: 66,
    _PASCAL: 256,
    __PRDL3_V1: 4,
}
```

I added my color mappings:
```
def create_prdl3_v1_colormap():
    colormap = np.zeros((256, 3), dtype=np.uint8)
    colormap[0] = [0, 0, 0] # void/other
    colormap[1] = [244, 66, 226] # chair
    colormap[2] = [66, 134, 244] # person
    colormap[3] = [66, 244,  95] # table
    return colormap
```

I added the lookup:
```
def create_label_colormap(dataset=_PASCAL):
  if dataset == _ADE20K:
    return create_ade20k_label_colormap()
  # ... etc
  elif dataset == __PRDL3_V1:
    return create_prdl3_v1_colormap()
  # ... etc
```

#### utils > train_utils.py

I added some weighting for class imbalance:
```
    # removed
    # not_ignore_mask = tf.to_float(tf.not_equal(scaled_labels,
    #                                            ignore_label)) * loss_weight

    # weights must be tuned
    ignore_weight = 0
    label0_weight = 1
    label1_weight = 15
    label2_weight = 10
    label3_weight = 15

    not_ignore_mask = \
    tf.to_float(tf.equal(scaled_labels, 0)) * label0_weight + \
    tf.to_float(tf.equal(scaled_labels, 1)) * label1_weight + \
    tf.to_float(tf.equal(scaled_labels, 2)) * label2_weight + \
    tf.to_float(tf.equal(scaled_labels, 3)) * label3_weight + \
    tf.to_float(tf.equal(scaled_labels, ignore_label)) * ignore_weight 
```

and I modified the exclude_list
```
exclude_list = ['global_step','logits']
```

#### datasets > dataset_generator.py

I registered the dataset:
```
_ade20k_PRDL3_v1_SEG_INFORMATION = DatasetDescriptor(
    splits_to_sizes={
        'train': 7009, 
        'trainval': 8761,
        'val': 1752,
    },
    num_classes=4, # 3 classes + 1 background = 4 classes
    ignore_label=255, # ignore nothing
)

_DATASETS_INFORMATION = {
    'cityscapes': _CITYSCAPES_INFORMATION,
    'pascal_voc_seg': _PASCAL_VOC_SEG_INFORMATION,
    'ade20k': _ADE20K_INFORMATION,
    'ade20k_PRDL3_v1': _ade20k_PRDL3_v1_SEG_INFORMATION,
}
```

#### My training script, deeplab > train-PRDL3_v1.sh

```
cd ..
# Set up the working environment.
CURRENT_DIR=$(pwd)
WORK_DIR="${CURRENT_DIR}/deeplab"
DATASET_DIR="datasets"

# Set up the working directories.
PDRL3_FOLDER="ade20k_PRDL3_v1"
EXP_FOLDER="exp/train_on_trainval_set"
INIT_FOLDER="${WORK_DIR}/${DATASET_DIR}/${PDRL3_FOLDER}/${EXP_FOLDER}/init_models"
TRAIN_LOGDIR="${WORK_DIR}/${DATASET_DIR}/${PDRL3_FOLDER}/${EXP_FOLDER}/train"
DATASET="${WORK_DIR}/${DATASET_DIR}/${PDRL3_FOLDER}/tfrecord"

mkdir -p "${WORK_DIR}/${DATASET_DIR}/${PDRL3_FOLDER}/exp"
mkdir -p "${TRAIN_LOGDIR}"

NUM_ITERATIONS=10
python "${WORK_DIR}"/train.py \
  --logtostderr \
  --train_split="train" \
  --model_variant="xception_65" \
  --atrous_rates=6 \
  --atrous_rates=12 \
  --atrous_rates=18 \
  --output_stride=16 \
  --decoder_output_stride=4 \
  --train_crop_size=481 \
  --train_crop_size=641 \
  --train_batch_size=2 \
  --training_number_of_steps="${NUM_ITERATIONS}" \
  --base_learning_rate=1e-3\
  --fine_tune_batch_norm=false \
  --initialize_last_layer=false \
  --last_layers_contain_logits_only=true \
  --tf_initial_checkpoint="${INIT_FOLDER}/deeplabv3_xception_ade20k_train/model.ckpt" \
  --train_logdir="${TRAIN_LOGDIR}" \
  --dataset_dir="${DATASET}"
```

#### My visualization script, deeplab > vis-PRDL3_v1.sh

```
cd ..
# Set up the working environment.
CURRENT_DIR=$(pwd)
WORK_DIR="${CURRENT_DIR}/deeplab"
DATASET_DIR="datasets"

# Set up the working directories.
PDRL3_FOLDER="ade20k_PRDL3_v1"
EXP_FOLDER="exp/train_on_trainval_set"
EVAL_LOGDIR="${WORK_DIR}/${DATASET_DIR}/${PDRL3_FOLDER}/${EXP_FOLDER}/eval"
VIS_LOGDIR="${WORK_DIR}/${DATASET_DIR}/${PDRL3_FOLDER}/${EXP_FOLDER}/vis"
DATASET="${WORK_DIR}/${DATASET_DIR}/${PDRL3_FOLDER}/tfrecord"
CHECKPOINT_DIR="${WORK_DIR}/datasets/ade20k_PRDL3_v1/exp/train_on_trainval_set/train"

mkdir -p "${EVAL_LOGDIR}"
mkdir -p "${VIS_LOGDIR}"


python "${WORK_DIR}"/vis.py \
--logtostderr \
--vis_split="val" \
--model_variant="xception_65" \
--atrous_rates=6 \
--atrous_rates=12 \
--atrous_rates=18 \
--output_stride=16 \
--decoder_output_stride=4 \
--fine_tune_batch_norm=false \
--vis_crop_size=481 \
--vis_crop_size=641 \
--checkpoint_dir="${CHECKPOINT_DIR}" \
--vis_logdir="${VIS_LOGDIR}" \
--dataset_dir="${DATASET}" \
--max_number_of_iterations=1 \
--colormap_type="ade20k_PRDL3_v1" \
--dataset="ade20k_PRDL3_v1"
```

#### The error

When I run the training script, a checkpoint is produced as expected. However, when running the visualization script, I get the following error:

```
Caused by op 'save/Assign_71', defined at:
  File "C:/Users/knk689/senior_design/tf_models/research/deeplab/vis.py", line 313, in <module>
    tf.app.run()
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\platform\app.py", line 125, in run
    _sys.exit(main(argv))
  File "C:/Users/knk689/senior_design/tf_models/research/deeplab/vis.py", line 286, in main
    session_creator=session_creator, hooks=None) as sess:
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\monitored_session.py", line 921, in __init__
    stop_grace_period_secs=stop_grace_period_secs)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\monitored_session.py", line 643, in __init__
    self._sess = _RecoverableSession(self._coordinated_creator)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\monitored_session.py", line 1107, in __init__
    _WrappedSession.__init__(self, self._create_session())
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\monitored_session.py", line 1112, in _create_session
    return self._sess_creator.create_session()
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\monitored_session.py", line 800, in create_session
    self.tf_sess = self._session_creator.create_session()
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\monitored_session.py", line 557, in create_session
    self._scaffold.finalize()
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\monitored_session.py", line 213, in finalize
    self._saver = training_saver._get_saver_or_default()  # pylint: disable=protected-access
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\saver.py", line 886, in _get_saver_or_default
    saver = Saver(sharded=True, allow_empty=True)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\saver.py", line 1102, in __init__
    self.build()
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\saver.py", line 1114, in build
    self._build(self._filename, build_save=True, build_restore=True)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\saver.py", line 1151, in _build
    build_save=build_save, build_restore=build_restore)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\saver.py", line 789, in _build_internal
    restore_sequentially, reshape)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\saver.py", line 459, in _AddShardedRestoreOps
    name="restore_shard"))
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\saver.py", line 428, in _AddRestoreOps
    assign_ops.append(saveable.restore(saveable_tensors, shapes))
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\training\saver.py", line 119, in restore
    self.op.get_shape().is_fully_defined())
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\ops\state_ops.py", line 221, in assign
    validate_shape=validate_shape)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\ops\gen_state_ops.py", line 61, in assign
    use_locking=use_locking, name=name)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\util\deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\framework\ops.py", line 3274, in create_op
    op_def=op_def)
  File "C:\Users\knk689\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\framework\ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): Restoring from checkpoint failed. This is most likely due to a mismatch between the current graph and the graph from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Assign requires shapes of both tensors to match. lhs shape= [4] rhs shape= [21]
         [[node save/Assign_71 (defined at C:/Users/knk689/senior_design/tf_models/research/deeplab/vis.py:286)  = Assign[T=DT_FLOAT, _class=["loc:@logits/semantic/biases"], use_locking=true, validate_shape=true, _device="/job:localhost/replica:0/task:0/device:GPU:0"](logits/semantic/biases, save/RestoreV2/_143)]]
         [[{{node save/RestoreV2/_1068}} = _Send[T=DT_FLOAT, client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device_incarnation=1, tensor_name="edge_1074_save/RestoreV2", _device="/job:localhost/replica:0/task:0/device:CPU:0"](save/RestoreV2:534)]]
```

**Comment by josepqp** [comment]
could you add your vis-PRDL3_v1.sh, you copied exactly the code used in train-PRDL3_v1.sh

**Comment by kayvonkhosrowpour** [comment]
> 
> 
> could you add your vis-PRDL3_v1.sh, you copied exactly the code used in train-PRDL3_v1.sh

My apologies. I've updated the code snippet.

**Comment by kayvonkhosrowpour** [comment]
Wow. After hours of backtracking and staring at my code, I present to you a solution:

```
python train.py \
 # everything as before ... \
 --dataset="NAME_OF_YOUR_DATASET_HERE"
```

I can't believe this has been tripping me up.

**Comment by ymodak** [comment]
Closing this issue since its resolved. Thanks!

**Comment by sidgan** [comment]
@kayvonkhosrowpour how did you arrive at

```
 ignore_weight = 0
 label0_weight = 1
 label1_weight = 15
 label2_weight = 10
 label3_weight = 15
```

for your dataset?

**Comment by kayvonkhosrowpour** [comment]
> @kayvonkhosrowpour how did you arrive at
> 
> ```
>  ignore_weight = 0
>  label0_weight = 1
>  label1_weight = 15
>  label2_weight = 10
>  label3_weight = 15
> ```
> 
> for your dataset?

They are hyperparameters that must be tuned. The ones you see are arbitrary.
